src/ngraph/pattern/matcher.hpp

```diff
@@ -124,6 +124,20 @@ namespace ngraph
             friend op::Label; //TODO: refine to match_class
 
         protected:
+            size_t add_node(std::shared_ptr<Node> node)
+            {
```

Comment:
It would be cleaner to have `add_node` just add the node and not return anything.
You could just set `watermark` to `m_matched_list.size()`.